drinkme/drinkme/Scenes/Auth/ViewModel/AuthViewModel.swift

```diff
@@ -0,0 +1,85 @@
+class AuthViewModel: AuthViewModelProtocol {
+    
+    // MARK: - Properties
+    var username: String = "";
+    var password: String = "";
+    
+    private(set) var isLoading: Bool = false {
+        didSet { onStateChanged?() }
+    }
+    
+    private(set) var errorMessage: String? {
+        didSet { onStateChanged?() }
+    }
+    
+    var onStateChanged: (() -> Void)?
+    
+    var onLoginSuccess: (() -> Void)?
+    
+    // MARK: - Dependencies
+    private let authService: AuthServiceProtocol
+    
+    // MARK: - Initializer
+    init(authService: AuthServiceProtocol) {
+        self.authService = authService
+    }
+    
+    // MARK: - Functions
+    func login() {
+        var (isValid, error) = validateUsername()
+        guard isValid else {
+            errorMessage = error
+            return
+        }
+        
+        (isValid, error) = validatePassword()
+        guard isValid else {
+            errorMessage = error
+            return
+        }
+        
+        isLoading = true
+        errorMessage = nil
+
+        authService.login(username: username, password: password, completion: { [weak self] result in
+            guard let self = self else { return }
+            
+            self.isLoading = false
+            if (result) {
+                self.onLoginSuccess?()
+            } else {
+                self.errorMessage = "Неверный пароль"
+            }
+        })
+    }
+    
+    func validateUsername() -> (Bool, String?) {
+        if (username.isEmpty) {
+            return (false, "Введите имя пользователя")
+        }
+        
+        if (!username.isAlphanumeric) {
```

Comment:
```suggestion
        if !username.isAlphanumeric {
```

Скобки можно не писать

Reply:
Не могу я адаптироваться к этому синаксису(( Скобки и `;` так и просятся